docs/rfcs/035-physical-replication.md

```diff
@@ -0,0 +1,197 @@
+replication a long time ago. However, we never properly wrote down all
+the decisions and assumptions, and in the last months when more users
+have started to use the feature, numerous issues have surfaced.
+
+This RFC documents the design decisions that have been made.
+
+## Summary
+
+PostgreSQL has a feature called streaming replication, where a replica
+streams WAL from the primary and continuously applies it. It is also
+known as "physical replication", to distinguish it from logical
+replication.  In PostgreSQL, a replica is initialized by taking a
+physical backup of the primary. In Neon, the replica is initialized
+from a slim "base backup" from the pageserver, just like a primary,
+and the primary and the replicas connect to the same pageserver,
+sharing the storage.
+
+There are two kinds of read-only replicas in Neon:
+- replicas that follow the primary, and
+- "static" replicas that are pinned at a particular LSN.
+
+A static replica is useful e.g. for performing time-travel query and
+running one-off slow queries without affecting the primary. A replica
+that follows the primary can be used e.g. to scale out read-only
+workloads.
+
+## Motivation
+
+Read-only replicas allow offloading read-only queries. It's useful for
+isolation, if you want to make sure that read-only queries don't
+affect the primary, and it's also an easy way to provide guaranteed
+read-only access to an application, without having to mess with access
+controls.
+
+## Non Goals (if relevant)
+
+This RFC is all about WAL-based *physical* replication. Logical
+replication is a different feature.
+
+Neon also has the capability to launch "static" read-only nodes which
+do not follow the primary, but are pinned to a particular LSN. They
+can be used for long-running one-off queries, or for Point-in-time
+query. They work similarly to read replicas that follow the primary,
+but some things are simpler: there are no concerns about cache
+invalidation when the data changes on the primary, or worrying about
+transactions that are in-progress on the primary.
+
+## Impacted components (e.g. pageserver, safekeeper, console, etc)
+
+- Control plane launches the replica
+- Replica Postgres instance connects to the safekeepers, to stream the WAL
+- The primary does not know about the standby, except for the hot standby feedback
+- The primary and replicas all connect to the same pageservers
+
+
+## Decisions, Issues
+
+### Cache invalidation in replica
+
+When a read replica follows the primary in PostgreSQL, it needs to
+stream all the WAL from the primary and apply all the records, to keep
+the local copy of the data consistent with the primary. In Neon, the
+replica can fetch the updated page versions from the pageserver, so
+it's not necessary to apply all the WAL. However, it needs to ensure
+that any pages that are currently in the Postgres buffer cache, or the
+Local File Cache, are either updated, or thrown away so that the next
```

Comment:
> or the Local File Cache, are either updated, or thrown away

Do we really actively throw away / evict pages from the LFC?

My expectation was that considering that RO advances its LSN, and that all pages in LFC are stamped with LSN (is it true?), we probably don't need to do anything proactively; old pages will be just evicted eventually anyway?

(I don't know much about LFC implementation, that's why this question)

Reply:
> > or the Local File Cache, are either updated, or thrown away
> 
> Do we really actively throw away / evict pages from the LFC?

Yes, see https://github.com/neondatabase/neon/blob/b7beaa0fd7a549634af04069d97dc7ef2d1aa7d1/pgxn/neon/pagestore_smgr.c#L3434.

This does not require an I/O, it just marks the page as unused in the LFC's in-memory hash table.

> My expectation was that considering that RO advances its LSN, and that all pages in LFC are stamped with LSN (is it true?), we probably don't need to do anything proactively; old pages will be just evicted eventually anyway?

If the page is updated by a WAL record, the version we have cached is outdated. We have to somehow make sure we don't try to use it.